### PR TITLE
support v2 composer

### LINKED ISSUE #13: Adds support for Composer v2
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes
| BC Break    | maybe ?

#### Summary

Composer v2 is behind the corner and we should be ready on the day 0, so we need updte this library ASAP.

This PR might be helpful:
https://github.com/Ocramius/PackageVersions/pull/133

### PR DESCRIPTION
**Pull request by samsonasik**
implements #13 fixes #13 
 

|    Q          |   A
|-------------- | ------
| BC Break      | no
| New Feature   | yes

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -1,23 +1,30 @@
 {
     "name": "laminas/laminas-dependency-plugin",
-    "description": "Replace zendframework and zfcampus packages with their Laminas Project equivalents.",
     "type": "composer-plugin",
+    "description": "Replace zendframework and zfcampus packages with their Laminas Project equivalents.",
     "license": "BSD-3-Clause",
-    "config": {
-        "sort-packages": true
-    },
```
Comment (michalbundyra) [inline composer.json:8]: please revert all unrelated changes in that file

File: src/DependencyRewriterPlugin.php [hunk 2]
```diff
@@ -71,6 +70,16 @@ public function activate(Composer $composer, IOInterface $io)
...
     }
 
+    public function deactivate(Composer $composer, IOInterface $io) : void
+    {
+        // Nothing to do here, as all features are provided through event listeners
+    }
+
+    public function uninstall(Composer $composer, IOInterface $io) : void
+    {
+        // Nothing to do here, as all features are provided through event listeners
```
Comment (boesing) [inline src/DependencyRewriterPlugin.php:80]: Actually, we are not even handling `uninstall` events with this plugin, thus we should not say assume that `uninstall` is handled by any listener.

File: src/ComposerV2Plugin.php [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+use Composer\Composer;
+use Composer\IO\IOInterface;
+use Composer\Plugin\PluginInterface;
+
+/**
+ * @internal do not rely on this: it is only here to declare the composer V2 API explicitly, even
+ *     in composer V1 context.
+ */
+interface ComposerV2Plugin extends PluginInterface
```
Comment (michalbundyra) [inline src/ComposerV2Plugin.php:21]: I don't think that we need that class

File: src/DependencyRewriterPlugin.php [hunk 4]
```diff
@@ -71,6 +70,16 @@ public function activate(Composer $composer, IOInterface $io)
         $this->output(sprintf('<info>Activating %s</info>', __CLASS__), IOInterface::DEBUG);
     }
 
+    public function deactivate(Composer $composer, IOInterface $io) : void
```
Comment (michalbundyra) [inline src/DependencyRewriterPlugin.php:73]: we can't have `: void` as we still on PH 5.6 and Composer v2 is gonna still support PHP 5.X

File: composer.json (hunk 1 above)
Comment (samsonasik) [reply composer.json:8]: reverted

File: src/ComposerV2Plugin.php (hunk 3 above)
Comment (samsonasik) [reply src/ComposerV2Plugin.php]: removed.

File: src/DependencyRewriterPlugin.php (hunk 4 above)
Comment (samsonasik) [reply src/DependencyRewriterPlugin.php]: `: void` removed.

File: src/DependencyRewriterPlugin.php (hunk 2 above)
Comment (samsonasik) [reply src/DependencyRewriterPlugin.php]: comment updated

**Comment by samsonasik** [comment]
@michalbundyra  I've added new tests for deactivate and uninstall

**Comment by michalbundyra** [comment]
@samsonasik check composer installation, because it looks like version 2 of composer is not installed. See: https://travis-ci.com/github/laminas/laminas-dependency-plugin/jobs/321131295#L284

**Comment by samsonasik** [comment]
it seems because of `dealerdirect/phpcodesniffer-composer-installer` still require v1, the travis config seems require --ignore-platform? tried locally but still got v1

**Comment by michalbundyra** [comment]
@samsonasik if i remember correctly it is because od [`minimum-stability`](https://getcomposer.org/doc/04-schema.md#minimum-stability).

**Comment by samsonasik** [comment]
I've updated with set minimum-stability and prefer-stable to travis config

**Comment by michalbundyra** [comment]
@samsonasik develop please, new feature.

**Comment by samsonasik** [comment]
updated target to develop branch

File: .travis.yml [hunk 5]
```diff
@@ -40,6 +40,8 @@ before_install:
   - if [[ $TEST_COVERAGE != 'true' ]]; then phpenv config-rm xdebug.ini || return 0 ; fi
 
 install:
+  - composer config minimum-stability dev
+  - composer config prefer-stable true
```
Comment (michalbundyra) [inline .travis.yml:44]: It cannot work, as composer v2 is not stable... yet.

Still tests does not run with composer v2, but we need run tests with composer v2 so we can make sure the feature  is implemented correctly.

**Comment by michalbundyra** [comment]
@samsonasik 

> updated target to develop branch

needs rebase, as it is based on master - see commits.
Also, squash, no point of all these changes and reverts and testing... 

But still it is not running tests on v2, so please investigate it on local, make sure it is using Composer v2 and then push the fix for tests.

File: .travis.yml (hunk 5 above)
Comment (samsonasik) [reply .travis.yml:44]: composer remove `dealerdirect/phpcodesniffer-composer-installer` on travis build?

Comment (samsonasik) [reply .travis.yml:44]: oh, I know, need update the composer exec as well in travis

Comment (samsonasik) [reply .travis.yml]: reverted.

**Comment by samsonasik** [comment]
rebased against develop branch

**Comment by samsonasik** [comment]
seems `2.0.x-dev` instead of `2.0@dev` checked at https://packagist.org/packages/composer/composer#dev-master

**Comment by samsonasik** [comment]
got the following error in php 7.2+ : 

```
There were 5 errors:

1) LaminasTest\DependencyPlugin\DependencyRewriterPluginTest::testOnPreDependenciesSolvingIgnoresNonInstallUpdateJobs
Prophecy\Exception\Doubler\MethodNotFoundException: Method `Double\Composer\Installer\InstallerEvent\P5::getRequest()` is not defined.

/Users/samsonasik/www/laminas-contribute/laminas-dependency-plugin/test/DependencyRewriterPluginTest.php:135

2) LaminasTest\DependencyPlugin\DependencyRewriterPluginTest::testOnPreDependenciesSolvingIgnoresJobsWithoutPackageNames
Prophecy\Exception\Doubler\MethodNotFoundException: Method `Double\Composer\Installer\InstallerEvent\P5::getRequest()` is not defined.

/Users/samsonasik/www/laminas-contribute/laminas-dependency-plugin/test/DependencyRewriterPluginTest.php:171

3) LaminasTest\DependencyPlugin\DependencyRewriterPluginTest::testOnPreDependenciesSolvingIgnoresNonZFPackages
Prophecy\Exception\Doubler\MethodNotFoundException: Method `Double\Composer\Installer\InstallerEvent\P5::getRequest()` is not defined.

/Users/samsonasik/www/laminas-contribute/laminas-dependency-plugin/test/DependencyRewriterPluginTest.php:207

4) LaminasTest\DependencyPlugin\DependencyRewriterPluginTest::testOnPreDependenciesSolvingIgnoresZFPackagesWithoutSubstitutions
Script phpunit --colors=always handling the test event returned with error code 2
Prophecy\Exception\Doubler\MethodNotFoundException: Method `Double\Composer\Installer\InstallerEvent\P5::getRequest()` is not defined.

/Users/samsonasik/www/laminas-contribute/laminas-dependency-plugin/test/DependencyRewriterPluginTest.php:246

5) LaminasTest\DependencyPlugin\DependencyRewriterPluginTest::testOnPreDependenciesSolvingReplacesZFPackagesWithSubstitutions
Error: Call to undefined method Composer\DependencyResolver\Request::install()

/Users/samsonasik/www/laminas-contribute/laminas-dependency-plugin/test/DependencyRewriterPluginTest.php:284
```

**Comment by samsonasik** [comment]
the code itself seems need update, there is no method `->getRequest()->getJob()` against `InstallerEvent` class.

**Comment by samsonasik** [comment]
@michalbundyra do you have reference for `getJob()` method in composer v2 ? It seems uses `$request->getRequires()`, but not sure how to get Request itself.

**Comment by boesing** [comment]
@samsonasik https://github.com/composer/composer/blob/master/UPGRADE-2.0.md

**Comment by samsonasik** [comment]
@boesing yes, I read that but didn't get the answer yet

**Comment by samsonasik** [comment]
I found in composer v2, the Request can be created via : 

```php
            $request = new Request(new LockArrayRepository());
            $jobs = $request->getRequires();
```

Still didn't find the relation to InstallerEvent.

File: src/DependencyRewriterPlugin.php [hunk 6]
```diff
@@ -115,8 +118,14 @@ public function onPreCommandRun(PreCommandRunEvent $event)
...
     {
         $this->output(sprintf('<info>In %s</info>', __METHOD__), IOInterface::DEBUG);
-        $request = $event->getRequest();
-        $jobs = $request->getJobs();
+        if (method_exists($event, 'getRequest')) {
+            $request = $event->getRequest();
+            $jobs = $request->getJobs();
+        } else {
+            $request = new Request(new LockArrayRepository());
+            $jobs = $request->getRequires();
```
Comment (boesing) [inline src/DependencyRewriterPlugin.php:126]: I'll check this out but just by seeing this, I cant think that this will return the same as in v1 `$requests->getJobs()` 🤔

File: test/DependencyRewriterPluginTest.php [hunk 7]
```diff
@@ -131,15 +133,17 @@ public function testOnPreDependenciesSolvingIgnoresNonInstallUpdateJobs()
         $this->activatePlugin($this->plugin);
 
         $event = $this->prophesize(InstallerEvent::class);
-        $request = $this->prophesize(Request::class);
-        $event->getRequest()->will([$request, 'reveal'])->shouldBeCalled();
+        if (method_exists(InstallerEvent::class, 'getRequest')) {
```
[the remark this answers is not in the document]
Comment (boesing) [reply test/DependencyRewriterPluginTest.php:136]: Maybe you create own tests for v1 and v2 as there are too many changes in the test if yo ask me

File: src/DependencyRewriterPlugin.php (hunk 6 above)
Comment (boesing) [reply src/DependencyRewriterPlugin.php:126]: Okay, I've dig a bit more into the magic of this library.

We have to totally rewrite it for composer 2.0 as its not about "detection" but to changing composer internals so it does continue with the replaced packages.

Comment (samsonasik) [reply src/DependencyRewriterPlugin.php:126]: thank you ;)

**Comment by michalbundyra** [comment]
@samsonasik what is the status of this PR? Are you going to carry on or @boesing is working on separate PR? Thanks

**Comment by samsonasik** [comment]
@michalbundyra I think I will let @boesing take care about it on separate PR, thank you ;)

**Comment by boesing** [comment]
@froschdesign Should the automation drop this PR instead of setting it to done? Shouldn't this only set to done if merged?
